polars/polars-lazy/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -7,43 +7,71 @@ use polars_time::prelude::*;
 
-            fn offset_fn(tu: TimeUnit) -> fn(&Duration, i64, Option<&Tz>) -> PolarsResult<i64> {
-                match tu {
-                    TimeUnit::Nanoseconds => Duration::add_ns,
-                    TimeUnit::Microseconds => Duration::add_us,
-                    TimeUnit::Milliseconds => Duration::add_ms,
-                }
-            }
+            let offset_fn = match tu {
+                TimeUnit::Nanoseconds => Duration::add_ns,
+                TimeUnit::Microseconds => Duration::add_us,
+                TimeUnit::Milliseconds => Duration::add_ms,
+            };
 
-            let out = match tz {
+            let tz_args = match tz {
                 #[cfg(feature = "timezones")]
-                Some(ref tz) => {
-                    let offset_fn = offset_fn(tu);
-                    ca.0.try_apply(|v| offset_fn(&offset, v, tz.parse::<Tz>().ok().as_ref()))
+                Some(ref tz) => tz.parse::<Tz>().ok(),
+                _ => None,
+            };
+
+            let out = match offsets.len() {
+                1 => {
+                    let offset = match offsets.get(0) {
+                        Some(offset) => Duration::parse(offset),
+                        _ => Duration::parse("0"),
+                    };
+                    ca.0.try_apply(|v| offset_fn(&offset, v, tz_args.as_ref()))
                 }
                 _ => {
-                    let offset_fn = offset_fn(tu);
-                    ca.0.try_apply(|v| offset_fn(&offset, v, None))
+                    let out = ca
+                        .into_iter()
+                        .zip(offsets.into_iter())
+                        .map(|(v, offset)| {
+                            let offset = match offset {
+                                Some(offset) => Duration::parse(offset),
+                                _ => Duration::parse("0"),
+                            };
+                            offset_fn(&offset, v.unwrap(), tz_args.as_ref()).unwrap()
+                        })
+                        .collect::<Vec<_>>();
+                    Ok(Int64Chunked::from_vec("", out))
                 }
             }?;
             // Sortedness may not be preserved when crossing daylight savings time boundaries
             // for calendar-aware durations.
             // Constant durations (e.g. 2 hours) always preserve sortedness.
-            preserve_sortedness =
-                tz.is_none() || tz.as_deref() == Some("UTC") || offset.is_constant_duration();
+            preserve_sortedness = match offsets.len() {
+                1 => {
+                    let offset = match offsets.get(0) {
+                        Some(offset) => Duration::parse(offset),
+                        _ => Duration::parse("0"),
+                    };
+                    tz.is_none() || tz.as_deref() == Some("UTC") || offset.is_constant_duration()
+                }
+                _ => tz.is_none() || tz.as_deref() == Some("UTC"),
```

Comment:
I might not be reading this right, but is the idea that if there are multiple offsets, then sortedness will be preseved in the tz-naive / tz-UTC case?

I'm not sure this is right - for example, if you start with
- [2020-01-01, 2020-01-02]

and offset by
- ['3d', '1d']

then the result will no longer be sorted

I think this may just need to be `false`?

If so, could we have a test check makes some assertion about the sortedness please?

Reply:
The idea was just me not fully understanding what the code was doing. ;_;

> I think this may just need to be false?

Makes sense now that you've explained it - thanks.

> If so, could we have a test check makes some assertion about the sortedness please?
 
Do you mean add to `test_datetime.py` similar to?

https://github.com/pola-rs/polars/blob/34d21463f7628471658b01c772db003647bdba40/py-polars/tests/unit/namespaces/test_datetime.py#L125

Reply:
yeah exactly, thanks!